File: bridges/python-ruby/tests/test-child.rb
# frozen_string_literal: true
#
# Ghost-Bridge test child (Ruby sidecar).
# Spawned by tests/test-client.py.
#
# Methods
# -------
#   echo(msg)            → msg
#   add(a, b)            → a + b
#   raise_error(message) → error response
#   echo_b64(data)       → base64(data)
#   slow(seconds)        → "done" after sleeping

require 'json'
require 'base64'

$stdout.sync = true

# Watchdog: exit when parent closes stdin
watchdog = Thread.new do
  # The main loop consumes stdin; this thread watches for the pipe itself
  # to become broken/closed from underneath us.
  sleep
rescue Exception
  exit 0
end

%w[INT TERM].each { |s| trap(s) { exit 0 } }

def write_response(hash)
  $stdout.print(hash.to_json + "\n")
  $stdout.flush
end

METHODS = {
  'echo' => lambda { |p|
    msg = p.fetch('msg') { raise ArgumentError, "missing param: msg" }
    msg
  },

  'add' => lambda { |p|
    a = p.fetch('a') { raise ArgumentError, "missing param: a" }
    b = p.fetch('b') { raise ArgumentError, "missing param: b" }
    a + b
  },

  'raise_error' => lambda { |p|
    message = p.fetch('message', 'intentional test error')
    raise RuntimeError, message
  },

  'echo_b64' => lambda { |p|
    data = p.fetch('data') { raise ArgumentError, "missing param: data" }
    Base64.strict_encode64(data.to_s)
  },

  'slow' => lambda { |p|
    seconds = p.fetch('seconds', 0.5).to_f
    sleep(seconds)
    'done'
  },
}.freeze

# ── Ready sentinel ────────────────────────────────────────────────────────────
write_response({ ready: true })

# ── Main loop ─────────────────────────────────────────────────────────────────
while (line = $stdin.gets)
  line = line.strip
  next if line.empty?

  begin
    req = JSON.parse(line)
  rescue JSON::ParserError => e
    $stderr.puts "[test-child] parse error: #{e.message}"
    next
  end

  req_id = req['id']
  method = req['method']
  params = req['params'] || {}

  handler = METHODS[method]
  if handler.nil?
    write_response({ id: req_id, error: { code: -32_601, message: "Method not found: #{method}" } })
    next
  end

  begin
    result = handler.call(params)
    write_response({ id: req_id, result: result })
  rescue StandardError => e
    write_response({ id: req_id, error: { code: -32_000, message: e.message } })
  end
end

exit 0

File: bridges/python-nodejs/tests/test-child.js
// tests/test-child.js - Node.js test sidecar for the Stitch Python→Node.js bridge.
//
// Spawned by tests/test-client.py and tests/python-nodejs_test.py.
//
// Methods
// -------
//   __ping__()            → { pong: true, pid }
//   echo({ msg })         → { msg }
//   add({ a, b })         → { sum: a + b }
//   raise_error({})       → throws Error("deliberate test error")
//   slow({ ms })          → sleeps ms milliseconds, then returns { done: true }
'use strict';

const readline = require('readline');

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Returns a Promise that resolves after `ms` milliseconds. */
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ---------------------------------------------------------------------------
// Handler registry
// ---------------------------------------------------------------------------
const handlers = {
  __ping__: async () => ({ pong: true, pid: process.pid }),

  echo: async ({ msg }) => {
    if (msg === undefined) throw new Error("missing param: msg");
    return { msg };
  },

  add: async ({ a, b }) => {
    if (a === undefined) throw new Error("missing param: a");
    if (b === undefined) throw new Error("missing param: b");
    return { sum: a + b };
  },

  raise_error: async () => {
    throw new Error("deliberate test error");
  },

  slow: async ({ ms = 100 }) => {
    await sleep(Number(ms));
    return { done: true };
  },
};

// ---------------------------------------------------------------------------
// Dispatch
// ---------------------------------------------------------------------------
async function dispatch(req) {
  const handler = handlers[req.method];
  if (!handler) {
    return {
      id: req.id,
      error: { code: -32601, message: `Method not found: ${req.method}` },
    };
  }
  try {
    const result = await handler(req.params || {});
    return { id: req.id, result };
  } catch (err) {
    return {
      id: req.id,
      error: { code: -32000, message: err.message || String(err) },
    };
  }
}

// ---------------------------------------------------------------------------
// I/O loop
// ---------------------------------------------------------------------------
const rl = readline.createInterface({ input: process.stdin });

// Ready signal – must be the very first line written.
process.stdout.write(JSON.stringify({ ready: true }) + '\n');

rl.on('line', async (line) => {
  const trimmed = line.trim();
  if (!trimmed) return;

  let req;
  try {
    req = JSON.parse(trimmed);
  } catch {
    return; // malformed JSON – no id to reply to
  }

  const resp = await dispatch(req);
  process.stdout.write(JSON.stringify(resp) + '\n');
});

// EOF watchdog – exit cleanly when the parent closes the pipe.
rl.on('close', () => process.exit(0));

// Signal traps.
process.on('SIGTERM', () => process.exit(0));
process.on('SIGINT',  () => process.exit(0));
